File: apps/comun/java/app/totalplayer/comun/Descargas.java
package app.totalplayer.comun;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import android.webkit.JavascriptInterface;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Guardar películas y episodios en el aparato, para verlos sin conexión.
 *
 * La interfaz es la misma web que en el navegador, y una web no puede
 * guardar dos gigas: lo que hay en un navegador es almacenamiento del sitio,
 * que el sistema borra cuando le hace falta espacio. Aquí sí se puede, así
 * que la web pregunta —«¿hay alguien que sepa descargar?»— y esta clase es la
 * que contesta que sí.
 *
 * El contrato con la web es a propósito de andar por casa: tres funciones que
 * reciben y devuelven texto. No es pereza, es que `@JavascriptInterface` no
 * sabe pasar objetos, y un puente que finge saber más de lo que sabe se rompe
 * en cuanto alguien lo toca. Está escrito igual en `components/tv/descargas.ts`.
 *
 * Los ficheros van a la carpeta privada de la aplicación: no hacen falta
 * permisos, no salen en la galería del televisor y se van con la aplicación
 * cuando se desinstala, que es lo que espera cualquiera.
 */
public class Descargas {

    private static final String TAG = "TPDescargas";
    /** Dónde se apunta lo que hay. El fichero se llama igual en las dos apps */
    private static final String CUADERNO = "tp_descargas";
    private static final String CLAVE = "lista";

    private final Context contexto;
    private final SharedPreferences cuaderno;
    /*
     * Una detrás de otra y no todas a la vez.
     *
     * Un televisor con wifi de casa y tres descargas en paralelo no baja tres
     * veces más rápido: baja lo mismo repartido, y mientras tanto el vídeo
     * que se está viendo se corta. En fila, la primera termina pronto y ya se
     * puede ver.
     */
    private final ExecutorService cola = Executors.newSingleThreadExecutor();
    /** Lo que se ha mandado parar mientras bajaba */
    private final Set<String> cancelados = new HashSet<>();
    private ServidorLocal servidor;

    public Descargas(Context contexto) {
        this.contexto = contexto.getApplicationContext();
        this.cuaderno = this.contexto.getSharedPreferences(CUADERNO, Context.MODE_PRIVATE);
        /*
         * Lo que quedara a medio bajar de la última vez, no está.
         *
         * Se apaga la tele con una descarga a la mitad y al volver a
         * encenderla el fichero está incompleto pero apuntado como si
         * estuviera bajando: sin esto, se queda «bajando 43%» para siempre.
         */
        repasarAlArrancar();
    }

    /* ---------- Lo que ve la web ---------- */

    /**
     * «Guárdame esto».
     *
     * El encargo trae identificador, nombre, carátula y la dirección del
     * vídeo ya resuelta contra el panel del proveedor: esa parte la hace la
     * web, que es la que tiene la sesión.
     */
    @JavascriptInterface
    public void bajar(String encargo) {
        try {
            final JSONObject e = new JSONObject(encargo);
            final String id = e.optString("id");
            final String origen = e.optString("url");
            if (id.isEmpty() || origen.isEmpty()) return;
            synchronized (this) {
                if (buscar(id) != null) return;  // ya está, o ya está bajando
                cancelados.remove(id);
                JSONObject fila = new JSONObject();
                fila.put("id", id);
                fila.put("nombre", e.optString("nombre"));
                fila.put("cartel", e.optString("cartel"));
                fila.put("estado", "bajando");
                fila.put("parte", 0);
                fila.put("bytes", 0);
                fila.put("origen", origen);
                anotar(fila);
            }
            cola.execute(new Runnable() {
                @Override
                public void run() {
                    traer(id, origen);
                }
            });
        } catch (Exception fallo) {
            Log.w(TAG, "encargo ilegible", fallo);
        }
    }

    /** «Quítame esto»: pare si está bajando, y borra lo que haya en disco. */
    @JavascriptInterface
    public void quitar(String id) {
        synchronized (this) {
            cancelados.add(id);
            borrar(id);
            File f = fichero(id);
            if (f.exists() && !f.delete()) Log.w(TAG, "no se ha podido borrar " + f);
            File medio = aMedias(id);
            if (medio.exists() && !medio.delete()) Log.w(TAG, "no se ha podido borrar " + medio);
        }
    }

    /** «¿Qué hay?». La web pregunta cada segundo y medio mientras algo baje. */
    @JavascriptInterface
    public String lista() {
        JSONArray fuera = new JSONArray();
        JSONArray dentro = leer();
        for (int i = 0; i < dentro.length(); i++) {
            JSONObject f = dentro.optJSONObject(i);
            if (f == null) continue;
            try {
                JSONObject copia = new JSONObject(f.toString());
                /* La dirección solo tiene sentido cuando el fichero está
                   entero: media película se ve como un vídeo roto */
                copia.put("url", "lista".equals(f.optString("estado")) ? donde(f.optString("id")) : "");
                copia.remove("origen");   // la web no tiene nada que hacer con ella
                fuera.put(copia);
            } catch (Exception ignorado) {
                /* Una fila ilegible no puede dejar sin lista a las demás */
            }
        }
        return fuera.toString();
    }

    /* ---------- Y lo mismo, para quien lo use desde Java ---------- */

    /*
     * La aplicación nativa de televisión no es una web: no hay puente ni
     * JavaScript, hay un reproductor que abre un fichero. Pero descargar es
     * exactamente lo mismo —la misma cola, el mismo `.medias`, el mismo
     * cuaderno—, así que se usa esta misma clase por la puerta de al lado.
     *
     * Lo que cambia es cómo se pregunta y cómo se ve: en Java se pasan y se
     * devuelven objetos, y el fichero se abre por su ruta y no por el
     * servidor local, que ahí no hace ninguna falta.
     */

    /** Una descarga, para quien la mira desde Java. */
    public static class Cosa {
        public String id = "";
        public String nombre = "";
        public String cartel = "";
        public String estado = "";
        public int parte;
        public long bytes;
        /** Dónde está el fichero, si ya está entero. Vacío si no. */
        public String ruta = "";
    }

    /** Encarga una descarga sin tener que montar el JSON a mano. */
    public void pedir(String id, String nombre, String cartel, String url) {
        try {
            JSONObject e = new JSONObject();
            e.put("id", id);
            e.put("nombre", nombre);
            e.put("cartel", cartel);
            e.put("url", url);
            bajar(e.toString());
        } catch (Exception fallo) {
            Log.w(TAG, "no se ha podido encargar " + id, fallo);
        }
    }

    /** Lo que hay guardado, de más reciente a más antiguo. */
    public java.util.List<Cosa> cosas() {
        java.util.List<Cosa> fuera = new java.util.ArrayList<>();
        JSONArray dentro = leer();
        for (int i = 0; i < dentro.length(); i++) {
            JSONObject f = dentro.optJSONObject(i);
            if (f == null) continue;
            Cosa c = new Cosa();
            c.id = f.optString("id");
            c.nombre = f.optString("nombre");
            c.cartel = f.optString("cartel");
            c.estado = f.optString("estado");
            c.parte = f.optInt("parte");
            c.bytes = f.optLong("bytes");
            c.ruta = "lista".equals(c.estado) ? rutaLocal(c.id) : "";
            fuera.add(c);
        }
        return fuera;
    }

    /** Cómo está una cosa concreta, o `null` si no está. */
    public Cosa comoVa(String id) {
        for (Cosa c : cosas()) if (c.id.equals(id)) return c;
        return null;
    }

    /**
     * Dónde está el fichero, ya en el aparato.
     *
     * Para el reproductor nativo esto es todo lo que hace falta: abre la
     * ruta y ya. El servidor local de aquí al lado solo existe porque un
     * WebView no puede abrir un `file://` desde una página en https.
     */
    public String rutaLocal(String id) {
        File f = fichero(id);
        return f.exists() ? f.getAbsolutePath() : "";
    }

    /** «1,4 GB». Vacío si todavía no se sabe cuánto ocupa. */
    public static String tamano(long bytes) {
        if (bytes <= 0) return "";
        if (bytes < 1024L * 1024) return Math.round(bytes / 1024f) + " KB";
        if (bytes < 1024L * 1024 * 1024) return Math.round(bytes / (1024f * 1024)) + " MB";
        return String.format(java.util.Locale.getDefault(), "%.1f GB", bytes / (1024f * 1024 * 1024));
    }

    /* ---------- Lo de dentro ---------- */

    /**
     * Baja el fichero, contando por dónde va.
     *
     * Se escribe en un `.medias` y solo al final se renombra: así, un corte
     * de luz deja basura reconocible y no una película que parece entera y
     * está partida por la mitad.
     */
    private void traer(String id, String origen) {
        HttpURLConnection con = null;
        InputStream entra = null;
        OutputStream sale = null;
        File destino = aMedias(id);
        try {
            URL url = new URL(origen);
            con = (HttpURLConnection) url.openConnection();
            con.setInstanceFollowRedirects(true);
            con.setConnectTimeout(20000);
            con.setReadTimeout(30000);
            /* Algunos paneles miran quién pide y contestan 403 a lo que no
               parece un reproductor */
            con.setRequestProperty("User-Agent", "TOTALplayer");
            con.connect();
            int codigo = con.getResponseCode();
            if (codigo < 200 || codigo >= 300) throw new Exception("respuesta " + codigo);

            long total = con.getContentLength();
            entra = con.getInputStream();
            sale = new FileOutputStream(destino);
            byte[] trozo = new byte[64 * 1024];
            long llevamos = 0;
            int ultimoAviso = -1;
            int leidos;
            while ((leidos = entra.read(trozo)) > 0) {
                if (cancelado(id)) {
                    cerrar(sale);
                    if (destino.exists() && !destino.delete()) Log.w(TAG, "sobra " + destino);
                    return;
                }
                sale.write(trozo, 0, leidos);
                llevamos += leidos;
                int parte = total > 0 ? (int) (llevamos * 100 / total) : 0;
                /* Se apunta cada punto porcentual y no cada trozo: escribir en
                   las preferencias sesenta veces por segundo es tocar el disco
                   sin necesidad y se nota en un aparato barato */
                if (parte != ultimoAviso) {
                    ultimoAviso = parte;
                    apuntarAvance(id, parte, llevamos);
                }
            }
            cerrar(sale);
            sale = null;
            if (cancelado(id)) {
                if (destino.exists() && !destino.delete()) Log.w(TAG, "sobra " + destino);
                return;
            }
            File entero = fichero(id);
            if (entero.exists() && !entero.delete()) Log.w(TAG, "no se ha podido reemplazar " + entero);
            if (!destino.renameTo(entero)) throw new Exception("no se ha podido cerrar el fichero");
            terminar(id, entero.length());
        } catch (Exception fallo) {
            Log.w(TAG, "no se ha podido bajar " + id, fallo);
            if (destino.exists() && !destino.delete()) Log.w(TAG, "sobra " + destino);
            if (!cancelado(id)) fallar(id);
        } finally {
            cerrar(sale);
            cerrar(entra);
            if (con != null) con.disconnect();
        }
    }

    private synchronized boolean cancelado(String id) {
        return cancelados.contains(id);
    }

    /** Dónde se ve, ya en el aparato. La sirve el servidor de aquí al lado. */
    private synchronized String donde(String id) {
        if (servidor == null || !servidor.enPie()) {
            servidor = new ServidorLocal(new ServidorLocal.Buscador() {
                @Override
                public File dame(String cual) {
                    File f = fichero(cual);
                    return f.exists() ? f : null;
                }
            });
            servidor.arrancar();
        }
        int puerto = servidor.puerto();
        if (puerto <= 0) return "";
        return "http://127.0.0.1:" + puerto + "/d/" + id;
    }

    private File carpeta() {
        File c = contexto.getExternalFilesDir("descargas");
        /* Sin tarjeta ni almacenamiento «externo» emulado —pasa en algún
           Fire TV viejo— se usa el interno, que siempre está */
        if (c == null) c = new File(contexto.getFilesDir(), "descargas");
        if (!c.exists() && !c.mkdirs()) Log.w(TAG, "no se ha podido crear " + c);
        return c;
    }

    /** Un nombre de fichero que no dependa de cómo se llame la película. */
    private File fichero(String id) {
        return new File(carpeta(), id.replaceAll("[^A-Za-z0-9_-]", "_") + ".dat");
    }

    private File aMedias(String id) {
        return new File(carpeta(), id.replaceAll("[^A-Za-z0-9_-]", "_") + ".medias");
    }

    /* ---------- El cuaderno ---------- */

    private synchronized JSONArray leer() {
        try {
            return new JSONArray(cuaderno.getString(CLAVE, "[]"));
        } catch (Exception fallo) {
            return new JSONArray();
        }
    }

    private synchronized void escribir(JSONArray lista) {
        cuaderno.edit().putString(CLAVE, lista.toString()).apply();
    }

    private synchronized JSONObject buscar(String id) {
        JSONArray lista = leer();
        for (int i = 0; i < lista.length(); i++) {
            JSONObject f = lista.optJSONObject(i);
            if (f != null && id.equals(f.optString("id"))) return f;
        }
        return null;
    }

    private synchronized void anotar(JSONObject fila) {
        JSONArray lista = leer();
        JSONArray nueva = new JSONArray();
        nueva.put(fila);                       // lo último encargado, arriba
        for (int i = 0; i < lista.length(); i++) nueva.put(lista.opt(i));
        escribir(nueva);
    }

    private synchronized void borrar(String id) {
        JSONArray lista = leer();
        JSONArray nueva = new JSONArray();
        for (int i = 0; i < lista.length(); i++) {
            JSONObject f = lista.optJSONObject(i);
            if (f != null && !id.equals(f.optString("id"))) nueva.put(f);
        }
        escribir(nueva);
    }

    private synchronized void cambiar(String id, String estado, int parte, long bytes) {
        JSONArray lista = leer();
        for (int i = 0; i < lista.length(); i++) {
            JSONObject f = lista.optJSONObject(i);
            if (f == null || !id.equals(f.optString("id"))) continue;
            try {
                if (estado != null) f.put("estado", estado);
                if (parte >= 0) f.put("parte", parte);
                if (bytes >= 0) f.put("bytes", bytes);
            } catch (Exception ignorado) {
                /* Un `put` de JSONObject no falla con estos tipos */
            }
            escribir(lista);
            return;
        }
    }

    private void apuntarAvance(String id, int parte, long bytes) {
        cambiar(id, null, parte, bytes);
    }

    private void terminar(String id, long bytes) {
        cambiar(id, "lista", 100, bytes);
    }

    private void fallar(String id) {
        cambiar(id, "fallo", -1, -1);
    }

    /**
     * Al arrancar, lo que quedó a medias queda como fallido.
     *
     * La cola de descargas no sobrevive a cerrar la aplicación, así que
     * cualquier cosa apuntada como «bajando» al arrancar es de la vez
     * anterior y no va a avanzar sola. Marcarla como fallida es lo honesto:
     * se ve, se sabe que ocupa disco, y se puede quitar o volver a pedir.
     */
    private void repasarAlArrancar() {
        JSONArray lista = leer();
        boolean tocado = false;
        for (int i = 0; i < lista.length(); i++) {
            JSONObject f = lista.optJSONObject(i);
            if (f == null || !"bajando".equals(f.optString("estado"))) continue;
            try {
                f.put("estado", "fallo");
                tocado = true;
            } catch (Exception ignorado) {
                /* No pasa con una cadena */
            }
        }
        if (tocado) escribir(lista);
    }

    private static void cerrar(Object c) {
        try {
            if (c instanceof InputStream) ((InputStream) c).close();
            if (c instanceof OutputStream) ((OutputStream) c).close();
        } catch (Exception ignorado) {
            /* Cerrar lo que ya está cerrado no es un problema de nadie */
        }
    }
}

File: apps/escritorio/src-tauri/src/descargas.rs
/*!
 * Guardar películas y episodios en el disco, para verlos sin conexión.
 *
 * La interfaz de este programa es la misma web que en el televisor, y una web
 * no puede guardar dos gigas: lo que hay en un navegador es almacenamiento
 * del sitio, que el sistema borra cuando le hace falta espacio. Aquí sí se
 * puede, así que la web pregunta —«¿hay alguien que sepa descargar?»— y esto
 * es lo que contesta que sí.
 *
 * El contrato con la web son tres funciones que reciben y devuelven texto, y
 * es igual de simple en la aplicación de Android por una razón que manda: el
 * puente de Android no sabe pasar objetos. Está escrito en
 * `components/tv/descargas.ts`, y la parte de Android en
 * `apps/comun/java/app/totalplayer/comun/Descargas.java`.
 *
 * Lo que sí cambia aquí es la forma de llamar: en Android la web llama a un
 * objeto de Java directamente y la respuesta es inmediata; en Tauri todo pasa
 * por `invoke`, que devuelve una promesa. Para que el contrato siga siendo el
 * mismo, el JavaScript que se inyecta guarda la última respuesta y `lista()`
 * devuelve esa. Se adapta el envoltorio al contrato, no al revés.
 */

use std::collections::HashSet;
use std::fs;
use std::io::{BufRead, BufReader, Read, Seek, SeekFrom, Write};
use std::net::{Ipv4Addr, TcpListener, TcpStream};
use std::path::PathBuf;
use std::sync::atomic::{AtomicU16, Ordering};
use std::sync::{Mutex, OnceLock};

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};

/// Una cosa guardada, o guardándose.
#[derive(Clone, Serialize, Deserialize)]
pub struct Fila {
    pub id: String,
    pub nombre: String,
    pub cartel: String,
    /// «bajando», «lista» o «fallo»
    pub estado: String,
    pub parte: u8,
    pub bytes: u64,
    /// De dónde se está bajando. No se le manda a la web: no tiene nada que
    /// hacer con ella y es una dirección con la sesión del proveedor dentro.
    #[serde(default)]
    pub origen: String,
}

/// Lo que se le manda a la web: lo de arriba, con `url` y sin `origen`.
#[derive(Serialize)]
struct Vista {
    id: String,
    nombre: String,
    cartel: String,
    estado: String,
    parte: u8,
    bytes: u64,
    url: String,
}

#[derive(Deserialize)]
struct Encargo {
    id: String,
    #[serde(default)]
    nombre: String,
    #[serde(default)]
    cartel: String,
    url: String,
}

fn cuaderno() -> &'static Mutex<Vec<Fila>> {
    static C: OnceLock<Mutex<Vec<Fila>>> = OnceLock::new();
    C.get_or_init(|| Mutex::new(Vec::new()))
}

/// Lo que se ha mandado parar mientras bajaba.
fn cancelados() -> &'static Mutex<HashSet<String>> {
    static C: OnceLock<Mutex<HashSet<String>>> = OnceLock::new();
    C.get_or_init(|| Mutex::new(HashSet::new()))
}

/// El puerto del servidor local. Cero mientras no esté abierto.
static PUERTO: AtomicU16 = AtomicU16::new(0);

fn carpeta(app: &AppHandle) -> PathBuf {
    let base = app
        .path()
        .app_data_dir()
        .unwrap_or_else(|_| PathBuf::from("."));
    let c = base.join("descargas");
    let _ = fs::create_dir_all(&c);
    c
}

/// Un nombre de fichero que no dependa de cómo se llame la película.
fn limpio(id: &str) -> String {
    id.chars()
        .map(|c| if c.is_ascii_alphanumeric() || c == '-' || c == '_' { c } else { '_' })
        .collect()
}

fn fichero(app: &AppHandle, id: &str) -> PathBuf {
    carpeta(app).join(format!("{}.dat", limpio(id)))
}

fn a_medias(app: &AppHandle, id: &str) -> PathBuf {
    carpeta(app).join(format!("{}.medias", limpio(id)))
}

fn indice(app: &AppHandle) -> PathBuf {
    carpeta(app).join("indice.json")
}

/// Se lee del disco una vez, al arrancar.
///
/// Lo que quedara a medio bajar de la vez anterior queda como fallido: la cola
/// de descargas no sobrevive a cerrar el programa, así que algo apuntado como
/// «bajando» al arrancar no va a avanzar solo. Decirlo es lo honesto — se ve,
/// se sabe que ocupa disco, y se puede quitar o volver a pedir.
pub fn al_arrancar(app: &AppHandle) {
    let leido: Vec<Fila> = fs::read_to_string(indice(app))
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .unwrap_or_default();
    let repasado = leido
        .into_iter()
        .map(|mut f| {
            if f.estado == "bajando" {
                f.estado = "fallo".into();
            }
            f
        })
        .collect();
    if let Ok(mut c) = cuaderno().lock() {
        *c = repasado;
    }
    guardar(app);
}

fn guardar(app: &AppHandle) {
    let copia = match cuaderno().lock() {
        Ok(c) => c.clone(),
        Err(_) => return,
    };
    if let Ok(texto) = serde_json::to_string(&copia) {
        let _ = fs::write(indice(app), texto);
    }
}

fn cambiar(app: &AppHandle, id: &str, estado: Option<&str>, parte: Option<u8>, bytes: Option<u64>) {
    if let Ok(mut c) = cuaderno().lock() {
        if let Some(f) = c.iter_mut().find(|f| f.id == id) {
            if let Some(e) = estado {
                f.estado = e.to_string();
            }
            if let Some(p) = parte {
                f.parte = p;
            }
            if let Some(b) = bytes {
                f.bytes = b;
            }
        }
    }
    guardar(app);
}

fn cancelado(id: &str) -> bool {
    cancelados().lock().map(|c| c.contains(id)).unwrap_or(false)
}

/* ---------- Lo que ve la web ---------- */

#[tauri::command]
pub fn tp_bajar(app: AppHandle, encargo: String) {
    let e: Encargo = match serde_json::from_str(&encargo) {
        Ok(e) => e,
        Err(_) => return,
    };
    if e.id.is_empty() || e.url.is_empty() {
        return;
    }
    {
        let mut c = match cuaderno().lock() {
            Ok(c) => c,
            Err(_) => return,
        };
        if c.iter().any(|f| f.id == e.id) {
            return; // ya está, o ya está bajando
        }
        // Lo último encargado, arriba
        c.insert(
            0,
            Fila {
                id: e.id.clone(),
                nombre: e.nombre.clone(),
                cartel: e.cartel.clone(),
                estado: "bajando".into(),
                parte: 0,
                bytes: 0,
                origen: e.url.clone(),
            },
        );
    }
    if let Ok(mut x) = cancelados().lock() {
        x.remove(&e.id);
    }
    guardar(&app);

    let hilo = app.clone();
    std::thread::spawn(move || traer(hilo, e.id, e.url));
}

#[tauri::command]
pub fn tp_quitar(app: AppHandle, id: String) {
    if let Ok(mut x) = cancelados().lock() {
        x.insert(id.clone());
    }
    if let Ok(mut c) = cuaderno().lock() {
        c.retain(|f| f.id != id);
    }
    let _ = fs::remove_file(fichero(&app, &id));
    let _ = fs::remove_file(a_medias(&app, &id));
    guardar(&app);
}

#[tauri::command]
pub fn tp_descargas(app: AppHandle) -> String {
    let puerto = servidor(&app);
    let c = match cuaderno().lock() {
        Ok(c) => c.clone(),
        Err(_) => return "[]".into(),
    };
    let fuera: Vec<Vista> = c
        .into_iter()
        .map(|f| {
            // La dirección solo tiene sentido cuando el fichero está entero:
            // media película se ve como un vídeo roto
            let url = if f.estado == "lista" && puerto > 0 {
                format!("http://127.0.0.1:{}/d/{}", puerto, f.id)
            } else {
                String::new()
            };
            Vista {
                id: f.id,
                nombre: f.nombre,
                cartel: f.cartel,
                estado: f.estado,
                parte: f.parte,
                bytes: f.bytes,
                url,
            }
        })
        .collect();
    serde_json::to_string(&fuera).unwrap_or_else(|_| "[]".into())
}

/* ---------- Bajarlo ---------- */

/// Se escribe en un `.medias` y solo al final se renombra: así, un corte de
/// luz deja basura reconocible y no una película que parece entera y está
/// partida por la mitad.
fn traer(app: AppHandle, id: String, origen: String) {
    let temporal = a_medias(&app, &id);
    let resultado = (|| -> Result<u64, String> {
        let res = ureq::AgentBuilder::new()
            .timeout_connect(std::time::Duration::from_secs(20))
            .build()
            .get(&origen)
            // Algunos paneles miran quién pide y contestan 403 a lo que no
            // parece un reproductor
            .set("User-Agent", "TOTALplayer")
            .call()
            .map_err(|e| e.to_string())?;
        let total: u64 = res
            .header("Content-Length")
            .and_then(|v| v.parse().ok())
            .unwrap_or(0);
        let mut entra = res.into_reader();
        let mut sale = fs::File::create(&temporal).map_err(|e| e.to_string())?;
        let mut trozo = vec![0u8; 64 * 1024];
        let mut llevamos: u64 = 0;
        let mut ultimo: i32 = -1;
        loop {
            if cancelado(&id) {
                return Err("cancelada".into());
            }
            let leidos = entra.read(&mut trozo).map_err(|e| e.to_string())?;
            if leidos == 0 {
                break;
            }
            sale.write_all(&trozo[..leidos]).map_err(|e| e.to_string())?;
            llevamos += leidos as u64;
            let parte = if total > 0 { (llevamos * 100 / total) as i32 } else { 0 };
            // Se apunta cada punto porcentual y no cada trozo: escribir el
            // índice sesenta veces por segundo es tocar el disco sin motivo
            if parte != ultimo {
                ultimo = parte;
                cambiar(&app, &id, None, Some(parte.clamp(0, 100) as u8), Some(llevamos));
            }
        }
        sale.flush().map_err(|e| e.to_string())?;
        drop(sale);
        let entero = fichero(&app, &id);
        let _ = fs::remove_file(&entero);
        fs::rename(&temporal, &entero).map_err(|e| e.to_string())?;
        Ok(fs::metadata(&entero).map(|m| m.len()).unwrap_or(llevamos))
    })();

    match resultado {
        Ok(bytes) => cambiar(&app, &id, Some("lista"), Some(100), Some(bytes)),
        Err(_) => {
            let _ = fs::remove_file(&temporal);
            if !cancelado(&id) {
                cambiar(&app, &id, Some("fallo"), None, None);
            }
        }
    }
}

/* ---------- Servirlo ---------- */

/**
 * Un servidor de tres líneas para servirle a la ventana lo que ya está en
 * disco.
 *
 * Suena a exageración y no lo es: la interfaz es una página en `https://`, y
 * desde una página así un `<video src="file:///...">` no carga —el navegador
 * lo bloquea, y hace bien—. Por `http://127.0.0.1` sí, porque el navegador lo
 * trata como sitio de confianza, y además entiende de rangos, que es lo que
 * permite avanzar y retroceder dentro de una película.
 *
 * No sale del equipo: escucha solo en la dirección de bucle local. Y solo
 * sirve lo que hay en la carpeta de descargas, porque lo que se pide no es
 * una ruta sino un identificador.
 */
fn servidor(app: &AppHandle) -> u16 {
    let ya = PUERTO.load(Ordering::Relaxed);
    if ya > 0 {
        return ya;
    }
    let puerta = match TcpListener::bind((Ipv4Addr::LOCALHOST, 0)) {
        Ok(p) => p,
        Err(_) => return 0,
    };
    let puerto = puerta.local_addr().map(|a| a.port()).unwrap_or(0);
    if puerto == 0 {
        return 0;
    }
    PUERTO.store(puerto, Ordering::Relaxed);
    let mio = app.clone();
    std::thread::spawn(move || {
        for quien in puerta.incoming() {
            let Ok(quien) = quien else { continue };
            let suyo = mio.clone();
            std::thread::spawn(move || {
                let _ = atender(&suyo, quien);
            });
        }
    });
    puerto
}

fn atender(app: &AppHandle, mut quien: TcpStream) -> std::io::Result<()> {
    let mut lineas = BufReader::new(quien.try_clone()?);
    let mut peticion = String::new();
    lineas.read_line(&mut peticion)?;
    let mut trozos = peticion.split_whitespace();
    let metodo = trozos.next().unwrap_or("");
    let camino = trozos.next().unwrap_or("").to_string();
    if metodo != "GET" {
        return seco(&mut quien, 405, "Method Not Allowed");
    }

    let mut desde: u64 = 0;
    let mut hasta: Option<u64> = None;
    loop {
        let mut linea = String::new();
        if lineas.read_line(&mut linea)? == 0 {
            break;
        }
        let linea = linea.trim_end().to_string();
        if linea.is_empty() {
            break;
        }
        let bajo = linea.to_ascii_lowercase();
        if let Some(rango) = bajo.strip_prefix("range:") {
            if let Some(bytes) = rango.trim().strip_prefix("bytes=") {
                let mut extremos = bytes.split('-');
                if let Some(a) = extremos.next() {
                    if let Ok(n) = a.trim().parse() {
                        desde = n;
                    }
                }
                if let Some(b) = extremos.next() {
                    if let Ok(n) = b.trim().parse() {
                        hasta = Some(n);
                    }
                }
            }
        }
    }

    let Some(id) = camino.strip_prefix("/d/") else {
        return seco(&mut quien, 404, "Not Found");
    };
    let ruta = fichero(app, id);
    if !ruta.exists() {
        return seco(&mut quien, 404, "Not Found");
    }
    let tamano = fs::metadata(&ruta)?.len();
    let hasta = hasta.filter(|h| *h < tamano).unwrap_or(tamano.saturating_sub(1));
    if desde >= tamano {
        return seco(&mut quien, 416, "Requested Range Not Satisfiable");
    }
    let cuanto = hasta - desde + 1;
    let por_trozos = desde > 0 || cuanto < tamano;

    let mut cabecera = String::new();
    cabecera.push_str(if por_trozos {
        "HTTP/1.1 206 Partial Content\r\n"
    } else {
        "HTTP/1.1 200 OK\r\n"
    });
    // El tipo se deja en genérico a propósito: lo que baja de un panel puede
    // ser mp4, mkv o ts, y el reproductor lo averigua mirando dentro mejor de
    // lo que lo adivinaríamos por la extensión
    cabecera.push_str("Content-Type: video/mp4\r\n");
    cabecera.push_str("Accept-Ranges: bytes\r\n");
    cabecera.push_str(&format!("Content-Length: {}\r\n", cuanto));
    if por_trozos {
        cabecera.push_str(&format!("Content-Range: bytes {}-{}/{}\r\n", desde, hasta, tamano));
    }
    // La página que lo pide está en otro origen —es la web, en https— así que
    // sin esto el navegador no la deja leerlo
    cabecera.push_str("Access-Control-Allow-Origin: *\r\n");
    cabecera.push_str("Connection: close\r\n\r\n");
    quien.write_all(cabecera.as_bytes())?;

    let mut lector = fs::File::open(&ruta)?;
    lector.seek(SeekFrom::Start(desde))?;
    let mut trozo = vec![0u8; 64 * 1024];
    let mut quedan = cuanto;
    while quedan > 0 {
        let piden = std::cmp::min(trozo.len() as u64, quedan) as usize;
        let leidos = lector.read(&mut trozo[..piden])?;
        if leidos == 0 {
            break;
        }
        // El reproductor corta la conexión en cuanto salta a otro punto del
        // vídeo: eso no es un error, es lo normal
        if quien.write_all(&trozo[..leidos]).is_err() {
            return Ok(());
        }
        quedan -= leidos as u64;
    }
    quien.flush()
}

fn seco(quien: &mut TcpStream, codigo: u16, texto: &str) -> std::io::Result<()> {
    let r = format!(
        "HTTP/1.1 {} {}\r\nContent-Length: 0\r\nConnection: close\r\n\r\n",
        codigo, texto
    );
    quien.write_all(r.as_bytes())
}

/**
 * El puente, tal y como lo espera la web.
 *
 * Se inyecta en la página igual que el atajo de recargar: es JavaScript a
 * secas que habla con el programa por `invoke`. Como el contrato es síncrono
 * —lo es porque el puente de Android no puede ser de otra manera—, aquí se
 * guarda la última respuesta y `lista()` devuelve esa; se pide la siguiente en
 * cuanto alguien pregunta, así que como mucho va un tirón por detrás. La web
 * pregunta cada segundo y medio mientras algo baja.
 */
pub const PUENTE: &str = r#"
(function () {
  function llamar(que, con) {
    try {
      return window.__TAURI_INTERNALS__.invoke(que, con || {});
    } catch (niIdea) {
      return Promise.reject(niIdea);
    }
  }
  var ultima = '[]';
  function refrescar() {
    llamar('tp_descargas').then(function (t) { ultima = t || '[]'; }).catch(function () {});
  }
  refrescar();
  window.TPDescargas = {
    bajar: function (encargo) { llamar('tp_bajar', { encargo: encargo }).then(refrescar, refrescar); },
    quitar: function (id) { llamar('tp_quitar', { id: id }).then(refrescar, refrescar); },
    lista: function () { refrescar(); return ultima; }
  };
})();
"#;
